File: from-2501-to-2600/2563-count-the-number-of-fair-pairs/solution.go
func countFairPairs(nums []int, lower int, upper int) int64 {
    sort.Ints(nums)
    n := len(nums)
    countLessThan := func(limit int) int64 {
        var result int64
        left, right := 0, n - 1
        for left < right {
            for nums[left] + nums[right] > limit && left < right {
                right -= 1
            }
            result += int64(right - left)
            left += 1
        }
        return result
    }
    return countLessThan(upper) - countLessThan(lower - 1)
}

File: from-2501-to-2600/2563-count-the-number-of-fair-pairs/solution.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        def bin_search(l, r, target) -> int:
            # Return largest i where nums[i] < target
            while l <= r:
                m = (l + r) // 2
                if nums[m] >= target:
                    r = m - 1
                else:
                    l = m + 1
            return r

        nums.sort()
        result = 0
        n = len(nums)
        for i in range(n):
            low = lower - nums[i]
            up = upper - nums[i]
            result += (
                bin_search(i + 1, n - 1, up + 1) - 
                bin_search(i + 1, n - 1, low)
            )
        return result
